[blog] Add Demucs ONNX export blog (#410)
mixxxdj/website: content/news/XXXX-XX-XX-gsoc2025-demucs-to-onnx-dhunstack.md

title: GSOC 2025 - Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format
authors: Anmol Mishra
status: draft
tags: gsoc, gsoc-2025, stems
comments: yes

Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*

> Mixxx currently supports playback of stem files, but does not support building stems from scratch. This project was a step towards supporting real time stem inference from within Mixxx. The key contributions of this project are -

Reviewer: ```suggestion
Mixxx 2.6 will support playback of stem files, but does not support building stems from scratch. This project was a step towards supporting real time stem inference from within Mixxx. The key contributions of this project are -
```
Stem files can be created from the original seperated tracks recorded in a studio (or from DAW) or build by stem seperation out of any track. You should be clear what this is about.
And add a link to https://www.stems-music.com/ that the reader can find out what the Stem file is.

Reviewer: The reader need more introduction to understand the following text. Explain especially what Demucs is.

- Preparing the Demucs code for ONNX export by rewriting non-exportable operations
- Validating all the modified parts numerically to ensure export of existing model weights without the need of retraining.
- Preparing scripts for exporting ONNX and ORT formats of the model.
- Example C++ scripts that serve as a minimal example of usage inside C++.
- Benchmarking the exported model for performance and separation quality.

## Why ONNX? Motivation and Challenges

Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute exported machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Exchange) has become a widely adopted serialization format, enabling language-independent deployment of machine learning models.

> While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these layers, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.

Reviewer: Explain what STFT and ISTFT are.
Also explain why you could reduce the fourier transformation from complex to the real part only.

| vocals.wav   | 7.42               | 7.41                  |
| **Overall**  | **10.70**          | **10.69**             |

*Table: SI-SDR (dB) comparison for each stem and overall, using torchmetrics. Results are shown for the native PyTorch model and the exported ONNX model running in C++.*

Now that we have a running platform independent high quality ONNX Demucs model that can utilize hardware acceleration and be deployed with C++, we plan to integrate this into Mixxx DJ for future.
We've prepared example scripts for running the exported Demucs model, which can be used following the instructions documented in our READMEs.

## Plans for future integration into Mixxx

Here's a list of all the PRs that document our incremental changes towards exporting Demucs:

- [feat] Add ONNX computation path — [mixxxdj/demucs#1](https://github.com/mixxxdj/demucs/pull/1)
- [feat] Add ONNX exportable STFT implementation — [mixxxdj/demucs#3](https://github.com/mixxxdj/demucs/pull/3)
- [feat] Add ONNX exportable ISTFT implementation — [mixxxdj/demucs#4](https://github.com/mixxxdj/demucs/pull/4)
- Add ONNX export instructions — [mixxxdj/demucs#5](https://github.com/mixxxdj/demucs/pull/5)

> 

Reviewer: You should mention upstreaming the changes to https://github.com/adefossez/demucs once you opened the PR

Also mention your upcoming talk at ADC2025

title: GSOC 2025 - Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format
authors: Anmol Mishra
status: draft
tags: gsoc, gsoc-2025, stems
math: yes
comments: yes

> Disclaimer: *This blog post primarily serves as the documentation for the [Google Summer of Code](https://summerofcode.withgoogle.com/programs/2025/projects/lRQpeA7K) 2025 project: "Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format".*

Reviewer 2: It would be great to have an eye catching vision statement at the very beginning. Something like: 

Imagine you could load a track and .....

> Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 

Reviewer 2: ```suggestion
Mixxx 2.6 will support playback of [stem files](https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 
```

Reviewer 2: A brief word explaining stem files would be nice to not loose readers in the first paragraph:  

support playback of [stem files](https://www.stems-music.com/) with contain separate channels for different instruments.

Reviewer 2: Let's use the Mixxx user perspective. Something like: 

.. which can obtained from from the original distribution, or if not available by one of the AI based splitter tools like ...

Reviewer 2: Demucs v4 can make use of its own paragraph. With an additional sentences how it works for non tech readers. 

Is "ONNX Runtime" the only option or shall we mention alternatives and than give a reason why ONNX Runtime works best for us.

Reviewer 2: There is a mix up with  "ONNX Runtime" and the ONNX format. Maybe we can mention runtimes that support that format.

> Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Antoine Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 

Reviewer: ```suggestion
Mixxx 2.6 will support playback of [stem files]( https://www.stems-music.com/), which can be created from the original track stems recorded in a studio (or DAW), or by using AI based stem separation models on the final mixed track. [Demucs v4](https://github.com/adefossez/demucs) is a state-of-the-art open-source music source separation model developed by Alexandre Défossez. While it provides exceptional quality in separating audio into stems, it is currently implemented in Python and therefore cannot be used directly in C++ applications or run efficiently on hardware accelerators through [ONNX Runtime](https://onnxruntime.ai/). 
```

> This project is a step towards supporting real time stem separation within Mixxx, by exporting the Python based Demucs model to [ONNX](https://onnx.ai/) (Open Neural Network Exchange format). The key contributions of this project are -

Reviewer 2: ```suggestion
This project is a step towards supporting real time stem separation within Mixxx, by exporting the Python based Demucs model to [ONNX](https://onnx.ai/) (Open Neural Network Exchange format). The key contributions of this project are:
```

- Preparing the Demucs code for ONNX export by rewriting non-exportable operations
- Validating all the modified parts with numerical tests to ensure export of existing model weights without the need for retraining.

> - Scripts for exporting ONNX and ORT formats of the model.

Reviewer 2: What is ORT? Can you mention it in the paragraph above?

Reviewer: ORT == ONNX Runtime: https://onnxruntime.ai/
Not self explaining for readers without background knowledge.

---

## Introduction

Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Exchange) has become a widely adopted serialization format, enabling language-independent deployment of machine learning models.

While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these layers, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.

> In this project, we successfully exported Demucs v4 as a fully self-contained ONNX model, including the STFT and ISTFT layers. This means the model can be used directly in any language that supports ONNX, greatly simplifying integration for audio developers and enabling broader adoption of source separation technology.

Reviewer 2: ... including the STFT and ISTFT layers

Can you explain this a bit more. Does ONNX now support that natively?

Reviewer: I think this sentence is clear and correct. The details of the implementation are described in the paragraphs below.

Reviewer 2: My issue is that I have no clue about layers in an ONNX model. This is not described anywhere else in this blog post. So you may describe it differently or just remove this detail..

Author: A `Layer` in the context of Deep Learning can be thought of as an abstracted operation.

DL evolved over time to have more complicated operations, so researchers started combining a standard set of fundamental operations together and created more abstract "layers", mainly for easier communication and concise description.
STFT and ISTFT can also be considered layers as they're a set of operations that transform the input in a precisely defined way. 

Large Deep Learning models are thus nothing but a set of these chained layers, however individual layers have become complex over time as they now combine other more simple layers. These more complex layers are also known as `blocks`.

In the text, I've used convolution and Transformer layers/blocks in the Demucs explanation.

But I'll remove this term here, will use "STFT and ISTFT operations" which might be better understood.

> A key part of our approach was to diagnose and address each export issue step by step, developing custom numerical tests to ensure that our ONNX-compatible replacements matched the original PyTorch calculations as closely as possible. This was crucial, as retraining the model was not a viable option: Demucs was originally trained on a large private dataset over weeks using multiple GPUs, and reproducing those results without the dataset and computational resources would be extremely difficult. By focusing on numerical fidelity, we ensured that the exported model retained the high performance of the original, without the need for retraining.

Reviewer 2: Why only "closely" and not "bit perfect"?

Author: Because PyTorch's STFT and ISTFT are implemented in C++ (and assembly). 
For ONNX export, I could only use the subset of PyTorch operations that are ONNX exportable, so the operations used can differ. Depending on how those operations are ultimately implemented, numerical differences could have occurred.
Which is why we did stringent tests, and benchmarks to evaluate the exported model.

---

## Understanding the Math Behind Audio

Before diving into transforms and neural networks, it helps to review how audio is represented mathematically and why **complex numbers** and **Fourier transforms** are so central to signal processing.

#### Real Numbers (Time-Domain Audio)

Digital audio is simply a sequence of **real numbers** — samples taken at regular time intervals.

Example: a 44.1 kHz stereo track stores **44,100 real values per second per channel**.  
Each value (e.g. `0.25`, `-0.67`) represents instantaneous amplitude (air pressure or voltage).

> #### Complex Numbers (Frequency Representation)

Reviewer 2: This is to fast. Something like this is missing: 

A complex number directed vector that can be the amplitude and a phase of a frequency in a track.

Reviewer 2: I wonder if that details contributes to the blog post. Maybe you can explain it implicit along with the FFT part below.

Reviewer: I think this helps to understand in the paragraphs below to understand why only the real part was necessary.

To analyze frequencies, we extend real numbers to **complex numbers**.

$$
z = a + bi
$$

Where:

- $a$: real part, $b$: imaginary part, $i^2 = -1$

Complex numbers are often represented in **polar form**:

$$
z = r e^{i\theta}
$$

Where:

- $r = \sqrt{a^2 + b^2}$ (magnitude)  

> - $\theta = \operatorname{atan2}(b, a)$ (phase angle)

Reviewer 2: I get this warning on GitHub: 
> The following macros are not allowed: operatorname

Not sure if this is a real issue.

Steps:

1. Split the signal into short, overlapping frames (e.g. 2048 samples).  
2. Apply a **window** (Hann/Hamming) to each frame.  
3. Compute the FFT of each windowed frame.

Mathematically:

$X(m,\omega) = \sum_n x[n] \, w[n - m] \, e^{-j \omega n}$

- $m$: frame index (time)  
- $\omega$: frequency bin  
- Output: **complex spectrogram** showing how frequency content evolves over time.

A spectrogram is a 2-D matrix (time × frequency) of complex values - often visualized by plotting magnitude in dB.

#### Inverse STFT (ISTFT)

To go back to the time domain:

1. Compute **inverse FFT** for each frame.  
2. Multiply by a synthesis window if required.  
3. **Overlap-add** frames to reconstruct the waveform.

Errors here cause audible artifacts (clicks, smearing, or phase shifts).

---

## Understanding HTDemucs and the Challenges with ONNX Export

To understand the export process, it’s important to look closely at how **HTDemucs (Hybrid Transformer Demucs)** is structured, and why its architecture introduces challenges when converting to the **ONNX** format.

> ![htdemucs]({static}/images/news/demucs-to-onnx/htdemucs.png)

Reviewer: For copyright reasons please add the information where you've taken the image from and as the file is taken from the Demucs project under MIT license, you also need to link the license file.

#### HTDemucs Explained

HTDemucs is a *hybrid* model — it combines two parallel processing branches:

##### **Time-Domain Branch**  

- Operates directly on the raw waveform (real-valued input).  
- Uses temporal convolutions and Transformer blocks to learn how each instrument evolves over time.

> ##### **Time-Frequency Domain Branch**  

Reviewer 2: I am missing a paragraph like this (my rookie understanding) 

The FFT is used to demangle the row of samples to the original bits of sound. Micro oscillations (sinusoid) like emitted of a certain bras sheet of a trumpet. The sum of these sinusoids can be converted back to the streams of sound samples. 

The Demux model takes all these resinoids and compares it to the training data and decide which of them belongs to which instrument. This way it can create one stream for each instrument.

Author: In this blog, I've tried to not talk much about the internal DL architecture, instead focusing only on the export related issues with Demucs. A discussion on the mechanisms of Demucs's layers might be too complex (and too much information) for this blog. 
I've referred to convolution and Transformer layers, whose internal mechanisms are well described mathematically but the math doesn't have an easy interpretation. Hence why the AI models are referred to as black boxes. 
I can link to the actual research paper though, if you suggest that'd be helpful - https://arxiv.org/pdf/2211.08553

Reviewer 2: Your new version reads already good. Thank you.

These two branches are then **fused** to produce the final stem-separated outputs (vocals, drums, bass, and others).

#### The ONNX Export Problem

The key difficulty in exporting HTDemucs lies in ONNX’s lack of support for complex tensors.

- In the **time-domain branch**, all operations are real-valued and export cleanly to ONNX.  
- In the **time-frequency branch**, two specific operations, **STFT** and **ISTFT**, involve complex numbers.

The rest of the model (all convolutional, transformer, and linear layers) operates purely on **real numbers**, and therefore poses **no export issue**.

ONNX, as of current opset versions, does not fully support:

- Complex-valued tensors as native types - [PyTorch Issue](https://github.com/pytorch/pytorch/issues/126972)
- Complex FFT operations (`torch.stft`, `torch.istft`) from PyTorch - [PyTorch Issue](https://github.com/pytorch/pytorch/issues/65666)

As a result, a direct `torch.onnx.export()` call on HTDemucs fails, since the exporter encounters unsupported complex operations.

---

## Our Solution

#### Real-Valued STFT and ISTFT Rewrites

> Because complex values appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** in a real-valued form.

Reviewer 2: This does not sound correct. Isn't the sine/coin notation just a different notation of the same complex value?

Reviewer: I think it is correct and uses the common mathematical terms. You could use the polar representation (magnitude and phase) instead of real-value + imaginary value. But to explain this work the later fits better.
sine/cosine are operations - not a representation.
Maybe a simple drawing like https://en.wikipedia.org/wiki/Complex_number#/media/File:Complex_number_illustration_modarg.svg would help the reader to understand the difference between magnitude, amplitude and phase?

Reviewer 2: I don't struggling to understand the complex number itself. 

For my understanding the sentences is misleading, because the complex-valued nature of the operation must still exists conceptually. It looks like the rewritten STFT avoids using complex tensors, by using a sinus and a cosinus temsor pair that is representing the same info. 

z=a+bi=reiθ=r(cosθ+isinθ)

Is that understanding correct?

Author: I see. Yeah the underlying math is same, but the implementation performs the operations separately for real part and imaginary part, so avoiding any complex tensor data type.
Would you think this statement might be a better fit?
`
Because complex tensors appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** using only real tensors.
`

Reviewer 2: Your explanation is very good, thank you. Unfortunately it is not integrated in the proposal. Here a try: 

```suggestion
Because complex tensors appear **only** in the initial STFT and final ISTFT layers, our solution was to **reimplement these operations** using separate real tensors for the real and the imagine part of the original complex values, while keeping the underlying math unchanged.
```

##### 1. **STFT Rewrite**

Instead of using `torch.stft()`, we expressed the Fourier transform as a set of **1-D convolutions** with precomputed sine and cosine kernels.

This allows us to compute:
$\text{Re}(X) = x * \cos(\omega)$, $\text{Im}(X) = -x * \sin(\omega)$

Thus, we can store the **real and imaginary parts** as separate real-valued tensors, fully ONNX-compatible.

##### 2. **ISTFT Rewrite**

Similarly, for inverse reconstruction:

- The original ISTFT combines complex values through real + imaginary synthesis.
- We reconstructed the time-domain signal by performing the same series of **overlap-add** and **cosine/sine inverse convolutions**, again using only real-valued tensors.

> By carefully ensuring numerical equivalence to PyTorch’s implementation, we achieved perfect parity (MSE < 1e-4) between the original and rewritten layers.

Reviewer 2: nice :-)

#### Unified Benchmarking: Timing and SI-SDR

To evaluate the exported ONNX model, we developed a unified benchmarking script (`benchmark.py`) that supports both PyTorch and ONNX/ORT backends. Key features:

- **Separation Backend:**
  - PyTorch: Runs separation using the original Demucs model.
  - ONNX: Runs separation by invoking the ONNX CLI tool via subprocess, matching the workflow of the provided example scripts.
- **Timing:**
  - The script records detailed timing for each track, printing and saving the results.
- **SI-SDR Evaluation:**
  - Computes the SI-SDR (Scale-Invariant Signal-to-Distortion Ratio) for each separated track using the `torchmetrics` library.
- **Output:**
  - Results are saved in timestamped folders, with timing and SI-SDR metrics output to JSON files named according to the backend (e.g., `results_onnx.json`).

#### How to Run the Benchmark

1. Place your test tracks in the input directory.
2. Run the benchmark script:

   ```python
   python benchmark.py --backend onnx --input_dir ./testfiles --output_dir ./outputs
   ```

3. Review the JSON results for timing and SI-SDR metrics.

## Benchmark Results

> The quality of the model is expected to be equal or slightly worse when exported to ONNX. While there are plenty of ways of measuring the benchmarks models (another blog post incoming), we've chosen to measure our models with `SI-SDR` metric, Scale Invariant Signal To Distortion Ratio, on the MusDB dataset. This is the standard metric on which researchers report their source separation model's performance.

Reviewer 2: Can you add other sentence with example db values? What would be perfect what is expected?

Reviewer: The numerical values for Demucs are in the table below. I don't think additional numbers will improve the text here.

Reviewer 2: I struggle to understand which value is better. Is 0 dB the optimum?

Author: @daschuer 
I can probably add a line here that higher dB values are better. 
This page is a great way to get familiarized with the SI-SDR metric - https://source-separation.github.io/tutorial/basics/evaluation.html
This is a table for SDR evaluation of Demucs compared to other models - https://github.com/facebookresearch/demucs?tab=readme-ov-file#comparison-of-accuracy

The evaluation of source separated tracks is an important topic in itself. The above URL I've mentioned talks about evaluation of ML models, but doesn't discuss the topic of quality assessment for exported models, which has multiple potential sources of quality degradation. Which is why I've added the comment for `(another blog post incoming)` in the text.

Reviewer 2: That works for me. Thank you.

Interesting links. You may add another sentence summarisimg your good explanations here.

Reviewer 2: Idea: for the subject part you may add two files for compassion python vs. c++.

- Validating all the modified parts with numerical tests to ensure export of existing model weights without the need for retraining.
- Scripts for exporting ONNX and ORT compatible formats of the model.
- Example scripts for deployment of ONNX model with C++ using ONNX Runtime.
- Benchmarking the exported model for performance and separation quality.
- Upstream PR with our modifications to the Demucs repository
- [Conference talk](https://conference.audio.dev/session/2025/converting-source-separation-models-to-onnx-for-real-time-usage-in-dj-software/) to be delivered at the ADC 2025

---

## Introduction

Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Exchange) has become a widely adopted serialization format, enabling language-independent deployment of machine learning models.

> While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like STFT and ISTFT, which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these operations, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.

Reviewer 3: Similarly to how you've done it when introducing the `ONNX` acronym, it would be great to express what STFT and ISTFT stand for, and also potentially link introduction that what they are (eg. [Wikipedia for STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform))

Edit: you do it later, but IMO it is great to give the full meaning the first time you use the acronym

## Plans for future integration into Mixxx

Here's a list of all the PRs that document our incremental changes towards exporting Demucs:

|     Task              |      PR Status      |                       PR Link                            |
|-----------------------|---------------------|----------------------------------------------------------|
| ONNX Computation Path |       Merged        |     [PR #1](https://github.com/mixxxdj/demucs/pull/1)    |
| STFT Rewrite          |       Merged        |     [PR #2](https://github.com/mixxxdj/demucs/pull/3)    |
| Inverse STFT Rewrite  |       Merged        |     [PR #4](https://github.com/mixxxdj/demucs/pull/4)    |
| ONNX Export Scripts   |       Merged        |     [PR #5](https://github.com/mixxxdj/demucs/pull/5)    |
| CI for model export   |       Merged        |     [PR #6](https://github.com/mixxxdj/demucs/pull/6)    |
| Example C++ Scripts   |      In Review      |     [PR #7](https://github.com/mixxxdj/demucs/pull/7)    |
| Benchmarking Scripts  |      In Review      |     [PR #9](https://github.com/mixxxdj/demucs/pull/8)    |

> We've raised a PR with all our Demucs changes to the upstream here - [ADD LINK].

Reviewer: Here is still a placeholder for the link!

- Validating all the modified parts with numerical tests to ensure export of existing model weights without the need for retraining.
- Scripts for exporting ONNX and ORT compatible formats of the model.
- Example scripts for deployment of ONNX model with C++ using ONNX Runtime.
- Benchmarking the exported model for performance and separation quality.
- Upstream PR with our modifications to the Demucs repository
- [Conference talk](https://conference.audio.dev/session/2025/converting-source-separation-models-to-onnx-for-real-time-usage-in-dj-software/) to be delivered at the ADC 2025

---

## Introduction

Source separation models are typically developed using Python-based libraries such as PyTorch or TensorFlow, which require a Python runtime to execute machine learning models. This dependency can make it challenging to use these models in environments where Python is not available or practical, such as audio plugins, which are often written in C++. To address this, ONNX (Open Neural Network Exchange) has become a widely adopted representation format, enabling language-independent deployment of machine learning models.

> While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like Short-time Fourier Transform ([STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform)) and Inverse Short-time Fourier Transform ([ISTFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform)), which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these operations, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.

Reviewer: ```suggestion
While both PyTorch and TensorFlow provide straightforward ONNX export for simple models, complex audio models like Demucs often present significant hurdles. This is primarily due to their use of complex tensors and custom implementations of operations like Short-time Fourier Transform ([STFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform)) and Inverse Short-time Fourier Transform ([ISTFT](https://en.wikipedia.org/wiki/Short-time_Fourier_transform#Inverse_STFT)), which are not natively supported by ONNX. As a result, previous attempts to export Demucs to ONNX have typically bypassed these operations, requiring developers to reimplement STFT/ISTFT in every target language, a process that is both error prone and time consuming.
```

We've prepared example scripts for running the exported Demucs model, which can be used following the instructions documented in our READMEs.

## Plans for future integration into Mixxx

Here's a list of all the PRs that document our incremental changes towards exporting Demucs:

|     Task              |      PR Status      |                       PR Link                            |
|-----------------------|---------------------|----------------------------------------------------------|
| ONNX Computation Path |       Merged        |     [PR #1](https://github.com/mixxxdj/demucs/pull/1)    |
| STFT Rewrite          |       Merged        |     [PR #2](https://github.com/mixxxdj/demucs/pull/3)    |
| Inverse STFT Rewrite  |       Merged        |     [PR #4](https://github.com/mixxxdj/demucs/pull/4)    |
| ONNX Export Scripts   |       Merged        |     [PR #5](https://github.com/mixxxdj/demucs/pull/5)    |
| CI for model export   |       Merged        |     [PR #6](https://github.com/mixxxdj/demucs/pull/6)    |
| Example C++ Scripts   |      In Review      |     [PR #7](https://github.com/mixxxdj/demucs/pull/7)    |

> | Benchmarking Scripts  |      In Review      |     [PR #9](https://github.com/mixxxdj/demucs/pull/8)    |

Reviewer: The benchmark PR #9 is missing and the link points to #8

---

## 🙏 Acknowledgements

This project would not have been possible without the extraordinary support and mentorship of [**Jörg**](https://github.com/JoergAtGithub), whose guidance, thoughtful reviews, and timely feedback shaped the project from start to finish. His expertise and encouragement were invaluable throughout every stage.

A huge thank you also to [**Antoine**](https://github.com/acolombier), who not only mentored me during the project but also integrated the exported ONNX model into his open-source tool [**Stemgen**](https://github.com/acolombier/stemgen), deploying it using **Rust** and **ONNX Runtime**. Seeing the model already being used in a real-world application is deeply rewarding.

Finally, I would like to thank the entire **Mixxx community** and **Google Summer of Code** program for providing me the opportunity to undertake this challenging endeavour and the right environment to support its successful completion.

## References

[1] A. Défossez, “Hybrid Transformers for Music Source Separation,” *arXiv preprint arXiv:2211.08553*, 2022.  

> Available: https://arxiv.org/abs/2211.08553

Reviewer: ```suggestion
Available: [https://arxiv.org/abs/2211.08553](https://arxiv.org/abs/2211.08553)
```

> title: GSOC 2025 - Converting Demucs v4 (Hybrid Transformer) AI model to ONNX format
> authors: Anmol Mishra
> tags: gsoc, gsoc-2025, stems

Reviewer: ```suggestion
authors: Anmol Mishra
status: draft
tags: gsoc, gsoc-2025, stems
```
I think the preview does not build, because this statement is missing
